src/service/git/git.ts

```diff
@@ -0,0 +1,296 @@
+     * @param cwd git repo
+     * @param ref ref from which the reachable commits are excluded
+     * @returns Array of parent commits
+     */
+    async getReachableParentCommits(cwd: string, ref: string): Promise<string[]>{
+        return (await this.git(cwd).raw("rev-list", "--parents", `${ref}..HEAD`))
+                .split(/\n/g)
+                .map(line => line.split(/ /g).slice(1))
+                .flat();
+    }
+
+    /**
+     * Equivalent to 'git pull --no-rebase url branch'
+     * @param cwd git repo in which the merge is to be performed
+     * @param repositoryUrl remote/url to pull from
+     * @param branch branch to pull from
+     */
+    async merge(cwd: string, repositoryUrl: string, branch: string) {
+        await this.git(cwd).pull(repositoryUrl, branch, ["--no-rebase"]);
+    }
+    
+    /**
+     * Return the sha-1 hash of the HEAD reference
+     * @param cwd git repo
+     * @returns sha string
+     */
+    async head(cwd: string): Promise<string> {
+        return await this.git(cwd).raw("show-ref", "--head", "-s", "/HEAD");
+    }
+     
+    /**
+     * Return the sha-1 hash of the given branch reference
+     * @param cwd git repo
+     * @param branch branch whose sha is to be returned
+     * @returns sha string
+     */
+    async sha(cwd: string, branch: string): Promise<string> {
+        return await this.git(cwd).raw("show-ref", "-s", `refs/remotes/origin/${branch}`);
+    }
+    
+    /**
+     * Rename the current branch
+     * @param cwd git repo
+     * @param newBranchName new branch name
+     */
+    async rename(cwd: string, newBranchName: string) {
+        await this.git(cwd).branch(["--move", newBranchName]);
+    }
+    
+    /**
+     * autosquash commits in on the given branch
+     * Equivalent to 'git rebase --quiet --autosquash branch'
+     * @param cwd git repo
+     * @param branch branch to be rebased
+     */
+    async rebase(cwd: string, branch: string) {
+        await this.git(cwd).rebase(["--quiet", "--autosquash", branch]);
+    }
+    
+    /**
+     * Perform a push
+     * Equivalent to 'git push --quiet origin branch' or if using force then
+     * equivalent to 'git push --quiet --force-with-lease origin branch'
+     * @param cwd git repo
+     * @param force whether to use force or not
+     * @param branch branch to push
+     */
+    async push(cwd: string, force: boolean, branch: string) {
+        const options = ["--quiet"];
+        if (force) {options.push("--force-with-lease");}
+        await this.git(cwd).push("origin", branch, options);
+    }
+
+    /**
+     * Check whether the given branch exists for the given repo and owner
+     * @param owner repo owner
+     * @param repo repo name
+     * @param branch branch that we need to check for existence
+     * @returns whether branch exists or not
+     */
+    async doesBranchExist(owner: string, repo: string, branch: string): Promise<boolean> {
```

Comment:
I'm thinking, should we split GitService into GitService and GithubAPIService

Reply:
I agree it will be easier to maintain too